src/Tools/ExternalAccess/FSharp/Internal/Editor/FSharpSynchronousIndentationService.cs

```diff
@@ -2,42 +2,61 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
 using System.Composition;
 using System.Threading;
 using Microsoft.CodeAnalysis.ExternalAccess.FSharp.Editor;
 using Microsoft.CodeAnalysis.Formatting;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Indentation;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.ExternalAccess.FSharp.Internal.Editor
 {
     [Shared]
     [ExportLanguageService(typeof(IIndentationService), LanguageNames.FSharp)]
     internal class FSharpSynchronousIndentationService : IIndentationService
     {
-        private readonly IFSharpSynchronousIndentationService _service;
+#pragma warning disable CS0618 // Type or member is obsolete
+        private readonly IFSharpSynchronousIndentationService? _legacyService;
+#pragma warning restore
+        private readonly IFSharpIndentationService? _service;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public FSharpSynchronousIndentationService(IFSharpSynchronousIndentationService service)
+        public FSharpSynchronousIndentationService(
+            [Import(AllowDefault = true)] IFSharpSynchronousIndentationService? legacyService,
+            [Import(AllowDefault = true)] IFSharpIndentationService? service)
         {
+            Contract.ThrowIfTrue(service == null && legacyService == null);
+
+            _legacyService = legacyService;
             _service = service;
         }
 
         public IndentationResult GetIndentation(Document document, int lineNumber, FormattingOptions.IndentStyle indentStyle, CancellationToken cancellationToken)
         {
-            var result = _service.GetDesiredIndentation(document, lineNumber, cancellationToken);
-            if (result.HasValue)
+            // all F# documents should have a file path
+            if (document.FilePath == null)
+            {
+                return default;
+            }
+
+            FSharpIndentationResult? result;
+            if (_service != null)
             {
-                return new IndentationResult(result.Value.BasePosition, result.Value.Offset);
+                var text = document.GetTextSynchronously(cancellationToken);
+                var options = document.GetOptionsAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+                var tabSize = options.GetOption(FormattingOptions.TabSize, LanguageNames.FSharp);
+                result = _service.GetDesiredIndentation(document.Project.LanguageServices, text, document.Id, document.FilePath, lineNumber, tabSize, indentStyle);
```

Comment:
i'm curious if we should have a special options struct particular to indentation.

Reply:
aside: i find it super icky that we're passing an ID + filepath... (i don't have a suggestion on what to do instead... but that's icky) :)

Reply:
> i'm curious if we should have a special options struct particular to indentation.

Possibly. The formatting options are quite complex. In any case, it wouldn't be public so we wouldn't expose that to F#.

Reply:
> aside: i find it super icky that we're passing an ID + filepath... (i don't have a suggestion on what to do instead... but that's icky) :)

Yeah, I'd rather not pass `DocumentId`, but F# is currently using it for some document and project based caching. Maybe another day we remove that.

Reply:
note: i don't mean passing the full formatting optinos.  i just mean a struct for things like "use tabs as spaces" "tab length" "indent style".  

actually... loooking at the sig now... how does F# know if "use tabs" is even on?  

--

I like an optiosn object as we can just add more options to it as necessary, instead of needing to grow this signature as a binary breaking change.  Thoughts?

Reply:
> note: i don't mean passing the full formatting optinos. i just mean a struct for things like "use tabs as spaces" "tab length" "indent style".

I know. I'm just saying that there are many formatting options used by various components and how exactly to split them to groups is not straightforward. But sure, I can see having IndentationOptions that contain the ones you listed. Will get there...

Reply:
> I like an optiosn object as we can just add more options to it as necessary, instead of needing to grow this signature as a binary breaking change. Thoughts?

Good point. I'll add `FSharpIndentationOptions`.

Reply:
> actually... loooking at the sig now... how does F# know if "use tabs" is even on?

IDK